rfcs/009 - Tree RFC.md

```diff
@@ -0,0 +1,110 @@
+
+## Author
+
+- David Ellis <david@alantechnologies.com>
+
+## Summary
+
+The ability to define and reason about data that has a tree-like structure is necessary to handle a host of different problems, from parsers and serializers, to manipulation of an HTML DOM or JSON object.
+
+This is an integral piece of any language. Most choose to allow recursive data types and then deal with the complexities around infinitely-defined types (one way or the other). For Alan, we instead define a new Tree type to handle this and provide a collection of tools to work with it.
+
+## Expected SemBDD Impact
+
+If Alan was post-1.0.0, this would be a minor update.
+
+## Proposal
+
+The `Tree` type is relatively straightforward:
+
+```ln
+type Tree<T> {
+  vals: Array<T>
+  parents: Array<int64>
+  children: Array<Array<int64>>
+}
+```
+
+where the values, parents and children relationships are tracked by three arrays. Each logical node is defined by matching indexes between them, with that being the node's ID, making the `Node` type straightforward:
+
+```ln
+type Node<T> {
+  id: int64
+  tree: Tree<T>
+}
+```
+
+This inversion of the relationship between Tree and Node avoids recursive representations while still allowing node-based manipulation a clean API.
+
+With basic functionality like tree construction, getting a root node, getting children, getting a node by value, etc, you could combine that with `@std/seq` to perform all of the tree manipulation you care for, but it is more interesting to provide methods similar to the array methods that can potentially perform certain operations in parallel.
+
+An equivalent to `map` for Tree makes sense, letting you transform the values of one tree into a new one, where the callback function could also use the parent and/or children to determine how to create the newly transformed node. The only question is whether or not it uses the same name of `map` or it gets a new name to be clearer?
+
+`reduce` is similarly interesting -- it could be parallelized by "dependency layer", all of the nodes with no children, then all of the nodes in the first layer above those, then the layer above that, etc, until you reach the root node. This means that serializing a data structure built on top of a Tree (like JSON, for instance) would be an automatically parallel operation and could therefore be more performant by default than other languages.
+
+`filter` could have two implementations, one that filters in a way that a failed filter node *and all of its children* are removed, and one where the failed nodes are deleted and their children (that pass the filter) are re-parented on the failed node's parent, where the latter would provide an array of potentially more than one tree if the root node is filtered away. What they should be called is not clear.
+
+`every` and `some` could simply be shorthand for something like `someTree.toNodeArray().every(...)` or just excluded instead.
```

Comment:
seems weird, probably just excluded

Reply:
The only one I'm not doing, because of offline discussions. Since we decided that mirroring the Array method names wherever possible was the right choice, excluding some array methods for no apparent reason doesn't seem right.